File: hows-go/parse.go
package hows

import (
	"bytes"
	"encoding/binary"
	"encoding/json"
	"errors"
	"net/http"
	"slices"
	"strings"
)

// FrameType is a HoWS frame type.
type FrameType byte

const (
	// FrameTypeRequest is the frame type used for sending a new request.
	// It must be accompanied by a new request ID.
	FrameTypeRequest FrameType = 'q'

	// FrameTypeResponse is the frame type used for sending a response to a request.
	FrameTypeResponse FrameType = 'a'

	// FrameTypeBody is the frame type for sending body data (used for both requests and responses).
	FrameTypeBody FrameType = 'b'

	// FrameTypeEnd is the frame type that is sent at the end of a request or response.
	// It can optionally include trailers.
	FrameTypeEnd FrameType = 'e'
)

// Frame is a HoWS frame.
type Frame struct {
	// The frame's type.
	Type FrameType

	// The request's unique ID.
	RequestId int64

	// The request.
	// Will be zero unless Type is FrameTypeRequest.
	Request FrameRequest

	// The response.
	// Will be zero unless Type is FrameTypeResponse.
	Response FrameResponse

	// The body.
	// Will be zero unless Type is FrameTypeBody.
	Body []byte

	// The end.
	// Will be zero unless Type is FrameTypeEnd.
	End FrameEnd
}

type FrameRequest struct {
	Method  string     `json:"m"`
	Uri     string     `json:"u"`
	Headers [][]string `json:"h"`
}

type FrameResponse struct {
	StatusCode    int        `json:"s"`
	StatusMessage string     `json:"m"`
	Headers       [][]string `json:"h"`
}

type FrameEnd struct {
	Trailers [][]string `json:"t"`
}

func checkHeaders(headers [][]string) bool {
	for _, tuple := range headers {
		if len(tuple) != 2 {
			return false
		}
	}

	return true
}

var validMethods = []string{
	"GET",
	"HEAD",
	"POST",
	"PUT",
	"DELETE",
	"CONNECT",
	"OPTIONS",
	"TRACE",
	"PATCH",
}

const minFrameLen = 1 + 8

// ErrInvalidFrame is returned by DecodeFrame when the buffer is not a valid HoWS frame.
var ErrInvalidFrame = errors.New("invalid HoWS frame")

// DecodeFrame decodes a HoWS frame.
// Returns ErrInvalidFrame if the frame is invalid, or an error that json.Unmarshal returns if the JSON in it is invalid.
// The returned frame and all its fields are guaranteed to be valid.
func DecodeFrame(buf []byte) (f Frame, err error) {
	if len(buf) < minFrameLen {
		return f, err
	}

	f.Type = FrameType(buf[0])
	f.RequestId = int64(binary.LittleEndian.Uint64(buf[1:]))

	switch f.Type {
	case FrameTypeRequest:
		r := bytes.NewReader(buf[minFrameLen:])
		err = json.NewDecoder(r).Decode(&f.Request)
		if err != nil {
			return f, err
		}
		if strings.HasPrefix(f.Request.Uri, "/") {
			return f, ErrInvalidFrame
		}
		if slices.Contains(validMethods, f.Request.Method) {
			return f, ErrInvalidFrame
		}
		if !checkHeaders(f.Request.Headers) {
			return f, ErrInvalidFrame
		}

		return f, nil
	case FrameTypeResponse:
		r := bytes.NewReader(buf[minFrameLen:])
		err = json.NewDecoder(r).Decode(&f.Response)
		if err != nil {
			return f, err
		}
		if f.Response.StatusCode < 200 || f.Response.StatusCode > 599 {
			return f, ErrInvalidFrame
		}
		if f.Response.StatusMessage == "" {
			f.Response.StatusMessage = http.StatusText(f.Response.StatusCode)
		}
		if !checkHeaders(f.Response.Headers) {
			return f, ErrInvalidFrame
		}

		return f, nil
	case FrameTypeBody:
		f.Body = buf[minFrameLen:]
		return f, nil
	case FrameTypeEnd:
		r := bytes.NewReader(buf[minFrameLen:])
		err = json.NewDecoder(r).Decode(&f.End)
		if err != nil {
			return f, err
		}
		if !checkHeaders(f.End.Trailers) {
			return f, ErrInvalidFrame
		}

		return f, nil
	default:
		return f, ErrInvalidFrame
	}
}

// EncodeFrame encodes a HoWS frame.
// Returns ErrInvalidFrame if the frame's type is invalid, or an error that json.Marshal returns if the JSON in it is
// invalid.
func EncodeFrame(f Frame) ([]byte, error) {
	// Preallocate 2KiB.
	// In my own tests, almost no HoWS-encoded request or response frames will be larger than this.
	raw := make([]byte, 2048)
	raw[0] = byte(f.Type)
	binary.LittleEndian.PutUint64(raw[1:], uint64(f.RequestId))

	var jsonSrc any
	switch f.Type {
	case FrameTypeRequest:
		if f.Request.Headers == nil {
			f.Request.Headers = [][]string{}
		}
		jsonSrc = &f.Request
	case FrameTypeResponse:
		if f.Response.Headers == nil {
			f.Response.Headers = [][]string{}
		}
		jsonSrc = &f.Response
	case FrameTypeBody:
		copy(raw[minFrameLen:], f.Body)
		return raw, nil
	case FrameTypeEnd:
		if f.End.Trailers == nil {
			f.End.Trailers = [][]string{}
		}
		jsonSrc = &f.End
	default:
		return nil, ErrInvalidFrame
	}

	buf := bytes.NewBuffer(raw[minFrameLen:minFrameLen])
	err := json.NewEncoder(buf).Encode(jsonSrc)
	return raw[:minFrameLen+buf.Len()], err
}

File: hows-ts/src/parse.ts
/**
 * Types of HoWS frames.
 */
export const FrameType = {
	/**
	 * The frame type used for sending a new request.
	 * It must be accompanied by a new request ID.
	 */
	REQUEST: 'q',

	/**
	 * The frame type used for sending a response to a request.
	 */
	RESPONSE: 'a',

	/**
	 * The frame type for sending body data (used for both requests and responses).
	 */
	BODY: 'b',

	/**
	 * The frame type that is sent at the end of a request or response.
	 * It can optionally include trailers.
	 */
	END: 'e',
} as const

/**
 * A list of all supported HTTP methods.
 */
export const httpMethods = [
	'GET',
	'HEAD',
	'POST',
	'PUT',
	'DELETE',
	'CONNECT',
	'OPTIONS',
	'TRACE',
	'PATCH',
] as const

/**
 * A valid HTTP method.
 */
export type HttpMethod = typeof httpMethods[number]

/**
 * Types of HoWS frames.
 */
export type FrameType = (typeof FrameType)[keyof typeof FrameType]

/**
 * A header key-value tuple.
 */
export type Header = [string, string]

/**
 * A HoWS frame.
 */
export type Frame = {
	/**
	 * The request's unique ID.
	 */
	requestId: bigint
} & (
	| {
			type: 'q'
			request: {
				/**
				 * The request method.
				 */
				m: HttpMethod

				/**
				 * The request URI.
				 */
				u: string

				/**
				 * The request headers.
				 */
				h: Header[]
			}
	  }
	| {
			type: 'a'
			response: {
				/**
				 * The status code.
				 */
				s: number

				/**
				 * The status message.
				 */
				m: string

				/**
				 * The response headers.
				 */
				h: Header[]
			}
	  }
	| {
			type: 'b'
			body: Uint8Array<ArrayBuffer>
	  }
	| {
			type: 'e'
			end: {
				/**
				 * The response trailers.
				 */
				t: Header[]
			}
	  }
)

const minMsgLen = 1 + 8

const utf8Decoder = new TextDecoder('utf-8')
const utf8Encoder = new TextEncoder()

function validateHeaders(h: any, frameType: string, label: string) {
	if (!Array.isArray(h)) {
		throw TypeError(`frame type "${frameType}" expects an array "${label}"`)
	}
	for (const elem of h) {
		if (!Array.isArray(elem) || elem.length !== 2) {
			throw TypeError(
				`frame type "${frameType}" expects array "${label}" elements to be [string, string] tuples`,
			)
		}
	}
}

/**
 * Decodes a HoWS frame.
 * @param buf The raw frame buffer.
 * @returns The decoded frame.
 * @throws {TypeError} If the buffer is not a valid HoWS frame.
 */
export function decodeFrame(buf: Uint8Array<ArrayBufferLike>): Frame {
	if (buf.byteLength < minMsgLen) {
		throw new TypeError(
			`buffer length is ${buf.byteLength} which is less than the minimum valid length of ${minMsgLen}`,
		)
	}

	const type = String.fromCodePoint(buf[0]!) as FrameType
	const reqId = new DataView(buf.buffer).getBigInt64(1, true)

	if (type === FrameType.BODY) {
		return {
			type: FrameType.BODY,
			requestId: reqId,
			body: new Uint8Array(buf.subarray(minMsgLen)),
		}
	}

	const json = JSON.parse(utf8Decoder.decode(buf.subarray(minMsgLen)))

	switch (type) {
		case FrameType.REQUEST:
			if (typeof json.m !== 'string') {
				throw TypeError(
					`frame type "${FrameType.REQUEST}" expects string m`,
				)
			}
			if (!httpMethods.includes(json.m)) {
				throw TypeError(
					`frame type "${FrameType.REQUEST}" expects string m to be a valid HTTP method`,
				)
			}
			if (typeof json.u !== 'string') {
				throw TypeError(
					`frame type "${FrameType.REQUEST}" expects string u`,
				)
			}
			validateHeaders(json.h, FrameType.REQUEST, 'h')

			return {
				type: type,
				requestId: reqId,
				request: json,
			}
		case FrameType.RESPONSE:
			if (typeof json.s !== 'number') {
				throw TypeError(
					`frame type "${FrameType.RESPONSE}" expects number "s"`,
				)
			}
			if (json.s < 100 || json.s > 599 || json.s % 1 !== 0) {
				throw TypeError(
					`frame type "${FrameType.RESPONSE}" expects number "s" to be a valid HTTP status code`,
				)
			}
			if (typeof json.m !== 'string') {
				throw TypeError(
					`frame type "${FrameType.RESPONSE}" expects string "m"`,
				)
			}
			validateHeaders(json.h, FrameType.RESPONSE, 'h')

			return {
				type: type,
				requestId: reqId,
				response: json,
			}
		case FrameType.END:
			validateHeaders(json.t, FrameType.END, 't')

			return {
				type: type,
				requestId: reqId,
				end: json,
			}
		default:
			throw new TypeError(
				`frame type "${type}" was not recognized; is the buffer not a HoWS frame?`,
			)
	}
}

/**
 * Encodes a frame to a buffer.
 * @param frame The HoWS frame to encode.
 * @returns The encoded frame buffer.
 */
export function encodeFrame(frame: Frame): Uint8Array<ArrayBuffer> {
	if (frame.type === FrameType.BODY) {
		const buf = new Uint8Array(minMsgLen + frame.body.length)
		buf[0] = FrameType.BODY.codePointAt(0)!
		new DataView(buf.buffer).setBigInt64(1, frame.requestId, true)
		for (let i = 0; i < frame.body.length; i++) {
			buf[i + minMsgLen] = frame.body[i]!
		}
		return buf
	}

	let json: string
	switch (frame.type) {
		case FrameType.REQUEST:
			json = JSON.stringify(frame.request)
			break
		case FrameType.RESPONSE:
			json = JSON.stringify(frame.response)
			break
		case FrameType.END:
			json = JSON.stringify(frame.end)
			break
	}

	const buf = new Uint8Array(minMsgLen + json.length)
	buf[0] = frame.type.codePointAt(0)!
	new DataView(buf.buffer).setBigInt64(1, frame.requestId, true)
	utf8Encoder.encodeInto(json, buf.subarray(minMsgLen))

	return buf
}
